rosbag2_transport/include/rosbag2_transport/visibility_control.hpp

```diff
@@ -0,0 +1,56 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* This header must be included by all rclcpp headers which declare symbols
+ * which are defined in the rclcpp library. When not building the rclcpp
+ * library, i.e. when using the headers in other package's code, the contents
+ * of this header change the visibility of certain symbols which the rclcpp
+ * library cannot have, but the consuming code must have inorder to link.
+ */
+
+#ifndef ROSBAG2_TRANSPORT__VISIBILITY_CONTROL_HPP_
+#define ROSBAG2_TRANSPORT__VISIBILITY_CONTROL_HPP_
+
+// This logic was borrowed (then namespaced) from the examples on the gcc wiki:
+//     https://gcc.gnu.org/wiki/Visibility
```

Comment:
over use of compiler macros should be avoided due to code complexity - is this to speedup compile time?

Reply:
also i just found this: seems you could reuse it? or are you purposefully avoiding that dependency?
https://github.com/ros2/rcutils/blob/master/include/rcutils/visibility_control.h

Reply:
> also i just found this: seems you could reuse it? or are you purposefully avoiding that dependency?

Unfortunately no, you cannot. Those files are actually different (controlled by different compiler definition) and that's required. The other option is to let CMake generate this file for each project, but personally I prefer static boilerplate to generated code when possible.

Reply:
> Perhaps you should read the linked wiki page?

I looked at the GNU wiki page and concluded it was for speedups... didn't realize it was required for Windows compatibility, sorry.

Reply:
Sorry you're right, the main benefit listed on that page is performance and DSO size, which it's real value from the perspective of GCC, but it's also required on Windows because they do not manage symbol exporting in DLL's for you, see the second about "Windows compatibility".

There are other resources about this:

- https://msdn.microsoft.com/en-us/library/a90k134d.aspx
- https://msdn.microsoft.com/en-us/library/3y1sfaz2.aspx

Reply:
I'll add some further explanations:

For ELF (Linux) and also OSX:

We actually don't use the speedup for gcc (yet) because the project is not compiled with `fvisibility=hidden`. In other words: After the first compiler pass this code is - at the moment - equivalent to code without these macros *for gcc* (and clang, because it mimics gcc in this regard). 

We should probably at some point compile with `fvisibility=hidden` (also to mimic Windows behaviour). This would then hide all unnecessary symbols from the `.so`-file, which decreases the size of the library as well as load times (because symbol allocations take less time). This could also allow for some further compiler optimizations but we don't compile with a high enough optimization level, yet.

On Windows (PE):

This code is required. The reason is that Windows forces you to explicitly declare all symbols that should be exported (i.e. visible to external dlls or executables consuming the dll). It also requires you to explicitly import all data. This is a requirement by the executable format and the easiest way is to do this via macros.

To conclude: You'll need at least a smaller version of this macro for Windows. It's easiest to take the whole macro and I believe we should at some point compile with `fvisibility=hidden` (and maybe `fvisibility-inlines-hidden`).